### PR TITLE
chore: Add check_uid constraint in DB

### PR DESCRIPTION
**Pull request by enricocolasante**
- Adding a `check_uid` function in the DB to verify that a varchar is a valid UID
- Adding a `uid_type` DB type to have columns that represents UID. It extends varchar and it checks that it is a valid UID.
- Adding constraints to `potentialduplicate` table to check that `uid`, `original` and `duplicate` columns are valid UID. `original` and `duplicate` are mapped as `UIDType` in Hibernate, so there is no way that we end up trying inserting an invalid UID in the database, while `uid` is mapped to `String` (it comes from `BaseIdentifiableObject`) and it can potentially be invalid. I added a test to check that DB will fail if we try to save a invalid UID in a column with the `check_uid` constraint.

### CODE REVIEW

**Review by gnespolino — APPROVED** [review]
good job

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/deduplication/DeduplicationServiceIntegrationTest.java [hunk 1]
```diff
@@ -110,6 +111,16 @@ void testAddPotentialDuplicate() throws PotentialDuplicateConflictException {
         deduplicationService.getPotentialDuplicateById(potentialDuplicate.getId()));
   }
 
+  @Test
+  void shouldFailWhenAddingAPotentialDuplicateWithInvalidUid() {
```
Comment (teleivo) [inline dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/deduplication/DeduplicationServiceIntegrationTest.java:115]: is it possible to add this test somewhere more related to uid/db things? I would not expect it here

File: dhis-2/dhis-support/dhis-support-db-migration/src/main/resources/org/hisp/dhis/db/migration/2.42/V2_42_30__Create_uid_check_constraint.sql [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+create or replace function check_uid(uid varchar) returns boolean
+    language plpgsql
+as
+$$
+begin
+     return uid ~ '^[A-Za-z]' and uid ~ '^[A-Za-z0-9]{11}$';
+end;
+$$;
+
+alter table potentialduplicate add constraint potentialduplicate_check_uid CHECK (check_uid(uid));
```
[the remark this answers is not in the document]
Comment (enricocolasante) [reply dhis-2/dhis-support/dhis-support-db-migration/src/main/resources/org/hisp/dhis/db/migration/2.42/V2_42_30__Create_uid_check_constraint.sql:13]: I went the other way :grimacing: 
Using the function to check that the type is valid. The function could still be useful and using the cast in the check is losing the context in the error of the actual constraint that is failing. Now, we get the specific constraint that is failing, not a generic, uid_type cannot be cast.

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/deduplication/DeduplicationServiceIntegrationTest.java (hunk 1 above)
Comment (enricocolasante) [reply dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/deduplication/DeduplicationServiceIntegrationTest.java]: I moved it under common, not sure it is the best place.
I accept suggestions.

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/deduplication/DeduplicationServiceIntegrationTest.java [hunk 3]
```diff
@@ -110,6 +111,16 @@ void testAddPotentialDuplicate() throws PotentialDuplicateConflictException {
...
+    PotentialDuplicate potentialDuplicate =
+        new PotentialDuplicate(TRACKED_ENTITY_A, TRACKED_ENTITY_B);
+    potentialDuplicate.setUid("INVALID");
+    assertThrows(
+        DataIntegrityViolationException.class,
```
[the remark this answers is not in the document]
Comment (enricocolasante) [reply dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/deduplication/DeduplicationServiceIntegrationTest.java]: Done, I am asserting on the constraint name on the exception message

**Review by stian-sandvold — APPROVED** [review]

**Pull request merged**